MLP.ipynb

```diff
@@ -1158,6 +1158,41 @@
     "plt.legend()\n",
     "plt.show()"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def find_best_model_configuration(param_grid, model, train_loader, val_loader, nn_network, device, nc):\n",
+    "    best_accuracy = 0\n",
+    "    best_params = None\n",
+    "\n",
+    "    for params in param_grid:\n",
+    "        print(f\"Testing configuration: {params}\")\n",
+    "        train_accuracies, train_losses, val_accuracies, val_losses = train_and_validate(\n",
+    "            model, train_loader, val_loader, nn_network, device, nc, learning_rate=params['learning_rate'], epochs=params['epochs'])\n",
+    "        # Get the best validation accuracy for the current parameter configuration\n",
+    "        max_val_accuracy = max(val_accuracies)\n",
+    "        \n",
+    "        # Update the best accuracy and parameters if the current model is better\n",
+    "        if max_val_accuracy > best_accuracy:\n",
+    "            best_accuracy = max_val_accuracy\n",
+    "            best_params = params\n",
+    "    return best_params, best_accuracy\n",
+    "\n",
+    "# Example usage:\n",
+    "param_grid = [\n",
+    "    {'epochs': 1, 'learning_rate': 0.01},\n",
+    "    {'epochs': 2, 'learning_rate': 0.001},\n",
```

Comment:
Here, In param grid, also we are hardcoding them only advantage is we can specify them together. 

Also, we need to reset the model and nn_network on every run, other wise the weights wont be re-initialised from the initial value.